content/blog/2024/09/usr-merge/index.md

```diff
@@ -0,0 +1,90 @@
+
+We started calling it the Chicken and the Egg problem.
+
+## Chicken and the egg
+
+The problem goes like this: we can update `eopkg` to provide more deterministic updates. In fact, Ermo has already done this for both the [Python 2](https://github.com/getsolus/eopkg/pull/68) and [Python 3](https://github.com/getsolus/eopkg/pull/70) versions of `eopkg`. But because updates are already non-deterministic, and we can't guarantee what point in time users are updating from, we can't rely on the update being present on user systems if we try to do the Usr Merge again; trying to update would lead to the same problem as before.
+
+And that holds true for any fix that involves a package update to solve this problem. No matter what package it is, or what the update does, the issue is always the same.
+
+## Cracking the egg
+
+We have come up with a way forward that should solve our little Chicken and Egg problem. We can't rely on a single update in a single package to solve the issue, but what we *can* do is update the pieces of software involved to write and look for marker files signaling that they are compatible with Usr Merge, and each other. In conjunction, we can ship a script to be run during the early boot process to do the actual merging, if and only if the marker files are in place. By having the process rely on these files and boot scripts instead of just happening with update installation, we can ensure that the merging process will only run when all of the pieces are in place, and not before. The package manager is taken out of the equation entirely. If you want to see what the script does for yourself, it can be found [here](https://github.com/getsolus/packages/blob/main/packages/u/usysconf-epoch/files/epoch.sh).
+
+### The plan
+
+The process goes like this:
+
+1) eopkg is updated with Usr-Merge support. It will write a marker file to indicate that the system can safely be merged.
+2) Our Usr Merge script runs during early boot and detects the marker file. The script proceeds to upgrade the machine safely, checking every file along the way with each step, only continuing if everything is correct.
+
+   The checking does mean that the script takes a while to run on less powerful systems; our testing in a Virtual Machine simulating really old hardware showed that the script took about 3 minutes to complete. Recent package updates include work to speed up the migration process significantly by essentially self-migrating, but it could still take a while.
+   
+   Because of the time the script could take to run, a message will be shown on screen so the user knows that something is happening. The script is robust enough to not destory your system if a reboot occurs while it is running, but doing a hard reboot does mean that the script has to start over on the next boot.
```

Comment:
```suggestion
   Because of the time the script could take to run, a message will be shown on screen so the user knows that something is happening. The script is robust enough to not destroy your system if a reboot occurs while it is running, but doing a hard reboot does mean that the script has to start over on the next boot.
```